clouddriver-appengine/src/main/groovy/com/netflix/spinnaker/clouddriver/appengine/deploy/ops/DeployAppEngineAtomicOperation.groovy

```diff
@@ -0,0 +1,140 @@
+import com.netflix.spinnaker.clouddriver.jobs.JobExecutor
+import com.netflix.spinnaker.clouddriver.jobs.JobRequest
+import com.netflix.spinnaker.clouddriver.jobs.JobStatus
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import org.springframework.beans.factory.annotation.Autowired
+
+class DeployAppEngineAtomicOperation implements AtomicOperation<DeploymentResult> {
+  private static final String BASE_PHASE = "DEPLOY"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  @Autowired
+  JobExecutor jobExecutor
+
+  DeployAppEngineDescription description
+
+  DeployAppEngineAtomicOperation(DeployAppEngineDescription description) {
+    this.description = description
+  }
+  /**
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "createServerGroup": { "application": "myapp", "stack": "stack", "freeFormDetails": "details", "repositoryUrl": "https://github.com/organization/project.git", "branch": "feature-branch", "credentials": "my-appengine-account", "appYamlPath": "path/to/app.yaml" } } ]' "http://localhost:7002/appengine/ops"
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "createServerGroup": { "application": "myapp", "stack": "stack", "freeFormDetails": "details", "repositoryUrl": "https://github.com/organization/project.git", "branch": "feature-branch", "credentials": "my-appengine-account", "appYamlPath": "path/to/app.yaml", "promote": true, "stopPreviousVersion": true } } ]' "http://localhost:7002/appengine/ops"
+   */
+  @Override
+  DeploymentResult operate(List priorOutputs) {
+    def directoryName = getLegalDirectoryName(description.repositoryUrl)
+
+    /*
+    * We can't allow concurrent deploy operations on the same local repository.
+    * If operation A checks out a new branch before operation B has run 'gcloud app deploy',
+    * operation B will deploy using that new branch's source files.
+    * */
+    AppEngineMutexRepository.atomicWrapper(directoryName, {
+      task.updateStatus BASE_PHASE, "Initializing creation of version..."
+      def result = new DeploymentResult()
+      def newVersionName = deploy(cloneOrUpdateLocalRepository(directoryName, 1))
+      result.serverGroupNames = [newVersionName]
+      result
+    })
+  }
+
+  String cloneOrUpdateLocalRepository(String directoryName, Integer retryCount) {
+    def repositoryUrl = description.repositoryUrl
+    def branch = description.branch
+    def directory = new File(directoryName)
+
+    try {
+      if (!directory.exists()) {
+        task.updateStatus BASE_PHASE, "Cloning repository $repositoryUrl into local directory..."
+        directory.mkdir()
+        runCommand(["git", "clone", repositoryUrl, directoryName])
+      }
+
+      task.updateStatus BASE_PHASE, "Checking out branch $branch and merging..."
+      runCommand(["git", "-C", directoryName, "fetch", "origin", branch])
+      runCommand(["git", "-C", directoryName, "checkout", branch])
```

Comment:
I could be wrong - but are you sure this branch will be created? I feel like that above `fetch` instead creates the branch `origin/$branch` rather than `$branch`.